azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java

```diff
@@ -6,550 +6,100 @@
-            info(FOUND_VALID_VALUE);
-            setter.accept(input);
-            return;
-        }
-
-        if (required) {
-            throw new MojoFailureException(String.format("invalid input: %s", input));
-        } else {
-            out.printf("The input is invalid. Use empty string.%n");
-            setter.accept("");
-        }
-    }
-
-    protected Scanner getScanner() {
-        return new Scanner(System.in, "UTF-8");
-    }
-
-    @Nullable
-    private String findElementInOptions(List<String> options, String item) {
-        return options.stream()
-            .filter(o -> o != null && o.equalsIgnoreCase(item))
-            .findFirst()
-            .orElse(null);
-    }
 
-    @Nullable
-    private List<String> getOptionsForUserPrompt(final String promptName) {
-        // HTTP Trigger
-        if ("authlevel".equalsIgnoreCase(promptName.trim())) {
-            return Arrays.asList("ANONYMOUS", "FUNCTION", "ADMIN");
-        }
-        // Cosmos DB Trigger
-        if ("createLeaseCollectionIfNotExists".equalsIgnoreCase(promptName.trim())) {
-            return Arrays.asList("true", "false");
-        }
-        return null;
-    }
+	@Parameter(defaultValue = "${project.basedir}", readonly = true, required = true)
+	protected File basedir;
+
+	@Parameter(defaultValue = "${project.compileSourceRoots}", readonly = true, required = true)
+	protected List<String> compileSourceRoots;
+
+	/**
+	 * Package name of the new function.
+	 *
+	 * @since 0.1.0
+	 */
+	@Parameter(property = "functions.package")
+	protected String functionPackageName;
+
+	/**
+	 * Name of the new function.
+	 *
+	 * @since 0.1.0
+	 */
+	@Parameter(property = "functions.name")
+	protected String functionName;
+
+	/**
+	 * Template for the new function
+	 *
+	 * @since 0.1.0
+	 */
+	@Parameter(property = "functions.template")
+	protected String functionTemplate;
+
+	public String getFunctionPackageName() {
+		return functionPackageName;
+	}
+
+	public String getFunctionName() {
+		return functionName;
+	}
+
+	public String getClassName() {
+		return getFunctionName().replace('-', '_');
+	}
+
+	public String getFunctionTemplate() {
+		return functionTemplate;
+	}
+
+	protected String getBasedir() {
+		return basedir.getAbsolutePath();
+	}
+
+	protected String getSourceRoot() {
+		return compileSourceRoots == null || compileSourceRoots.isEmpty()
+				? Paths.get(getBasedir(), "src", "main", "java").toString()
+				: compileSourceRoots.get(0);
+	}
+
+	protected void setFunctionPackageName(String functionPackageName) {
+		this.functionPackageName = StringUtils.lowerCase(functionPackageName);
+	}
+
+	protected void setFunctionName(String functionName) {
+		this.functionName = StringUtils.capitalize(functionName);
+	}
+
+	protected void setFunctionTemplate(String functionTemplate) {
+		this.functionTemplate = functionTemplate;
+	}
+
+	@Override
+	protected void doExecute() throws Exception {
+		AddFunctionHandler handler = new AddFunctionHandler();
```

Comment:
Mojo should do the job of asking for parameters and handler should only accept the inputs and generate function file

Reply:
good question, handler should keep some interfaces with the properties it requires.